sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/FailureSafeParser.scala

```diff
@@ -55,11 +56,15 @@ class FailureSafeParser[IN](
 
   def parse(input: IN): Iterator[InternalRow] = {
     try {
-     if (skipParsing) {
-       Iterator.single(InternalRow.empty)
-     } else {
-       rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
-     }
+      if (skipParsing) {
+        if (unparsedRecordIsNonEmpty(input)) {
+          Iterator.single(InternalRow.empty)
+        } else {
+          Iterator.empty
+        }
+      } else {
+        rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
+      }
```

Comment:
We need to modify this file? Since this is a issue in json stuffs, I think its better to handle this case in the json parser side. Can't we do handle this in the same way with CSV one?, e.g., 
https://github.com/apache/spark/blob/3a17c6a06bc4e8a6cae506a61d3d83696d378607/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala#L333

Reply:
One of my earlier revisions [worked in the way you suggest](https://github.com/sumitsu/spark/blob/13942b8da17d7e58e690b7a8b27c1a3fe20be135/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala#L132-L160) (if I've understood your point correctly); I changed it in order to avoid redundant empty-line filtering in the case where the full parser has to run anyway (i.e. where `skipParsing == false`).

What do you think? Is that a valid optimization, or is it better to do it on the JSON code as in 13942b8 to avoid changes to `FailureSafeParser`?

Reply:
I thought this: https://github.com/maropu/spark/commit/f4df9074619e6cafff679c5f22fa153727be1079
But, you should follow other guys who are familiar this part, @HyukjinKwon and @MaxGekk